x-pack/plugin/wildcard/src/main/java/org/elasticsearch/xpack/wildcard/mapper/NGramAutomaton.java

```diff
@@ -0,0 +1,374 @@
+            // later. In that case we just give up.
+            initialStates.clear();
+            states.clear();
+            return false;
+        }
+        if (offset == gramSize - 1) {
+            // We've walked deeply enough to find an initial state.
+            NGramState state = new NGramState(currentState, new String(codePoints, 0, gramSize - 1), true);
+            // Only add one copy of each state - if we've already seen this
+            // state just ignore it.
+            if (states.containsKey(state)) {
+                return true;
+            }
+            initialStates.add(state);
+            states.put(state, state);
+            return true;
+        }
+        // TODO build fewer of these
+        Transition transition = new Transition();
+        int totalLeavingState = source.initTransition(currentState, transition);
+        for (int currentLeavingState = 0; currentLeavingState < totalLeavingState; currentLeavingState++) {
+            source.getNextTransition(transition);
+            int min, max;
+            if (transition.max - transition.min >= maxExpand) {
+                // Consider this transition useless.
+                min = 0;
+                max = 0;
+            } else {
+                min = transition.min;
+                max = transition.max;
+            }
+            for (int c = min; c <= max; c++) {
+                codePoints[offset] = c;
+                if (!buildInitial(codePoints, offset + 1, transition.dest)) {
+                    return false;
+                }
+            }
+        }
+        return true;
+    }
+
+    private void traceRemainingStates() {
+        LinkedList<NGramState> leftToProcess = new LinkedList<NGramState>();
+        leftToProcess.addAll(initialStates);
+        int[] codePoint = new int[1];
+        int statesTraced = 0;
+        Transition transition = new Transition();
+        int currentTransitions = 0;
+        while (!leftToProcess.isEmpty()) {
+            if (statesTraced >= maxStatesTraced) {
+                throw new AutomatonTooComplexException();
+            }
+            statesTraced++;
+            NGramState from = leftToProcess.pop();
+            if (acceptStates.contains(from)) {
+                // Any transitions out of accept states aren't interesting for
+                // finding required ngrams
+                continue;
+            }
+            int totalLeavingState = source.initTransition(from.sourceState, transition);
+            if (currentTransitions >= maxTransitions) {
+                acceptStates.add(from);
+                continue;
+            }
+            for (int currentLeavingState = 0; currentLeavingState < totalLeavingState; currentLeavingState++) {
+                source.getNextTransition(transition);
+                int min, max;
+                if (transition.max - transition.min >= maxExpand) {
+                    // Consider this transition useless.
```

Comment:
It's probably worth refactoring the transition expansion logic into *something* eventually. And this comment isn't great. I did write it years ago, so that is 100% my fault.